### ISSUE TITLE
docker image is duplicating db files somehow

### ISSUE THREAD

**Issue by fgregg** [simonw/datasette#1836]
if you look into the docker image created by docker publish, the `datasette inspect` line is duplicating the db files.

here's the result of the inspect command:

<img width="490" alt="Screen Shot 2022-10-06 at 2 58 08 PM" src="https://user-images.githubusercontent.com/536941/194430909-9303ee1a-9bdd-4212-b54b-de28f43768d4.png">

**Comment by fgregg** [comment]
i thought it was maybe to do with reading through all the files, but that does not seem to be the case

if i make a little test file like:

```python
# test_read.py
import hashlib
import sys
import pathlib

HASH_BLOCK_SIZE = 1024 * 1024

def inspect_hash(path):
    """Calculate the hash of a database, efficiently."""
    m = hashlib.sha256()
    with path.open("rb") as fp:
        while True:
            data = fp.read(HASH_BLOCK_SIZE)
            if not data:
                break
            m.update(data)

    return m.hexdigest()

inspect_hash(pathlib.Path(sys.argv[1]))
```

then a line in the Dockerfile like

```docker
RUN python test_read.py nlrb.db && echo "[]" > /etc/inspect.json
```

just produes a layer of `3B`

**Comment by fgregg** [comment]
it's not that the inspect command is somehow changing the db files. if i set them to only read-only, the "inspect" layer still has the same very large size.

**Comment by fgregg** [comment]
okay, some progress!! running some sql against a database file causes that file to get duplicated even if it doesn't apparently change the file.

make a little test script like this:

```python
import sqlite3
import sys

conn = sqlite3.connect(sys.argv[1])
cur = conn.cursor()
cur.execute('select count(*) from filing')
print(cur.fetchone())
```

then 

```docker
RUN python test_sql.py nlrb.db
```

produced a layer that's the same size as `nlrb.db`!!

**Comment by fgregg** [comment]
fascinatingly, telling python to open sqlite in read only mode makes this layer have a size of 0

```python
# test_sql_ro.py
import sqlite3
import sys

db_name = sys.argv[1]
conn = sqlite3.connect(f'file:/app/{db_name}?mode=ro', uri=True)
cur = conn.cursor()
cur.execute('select count(*) from filing')
print(cur.fetchone())
```

that's quite weird because setting the file permissions to read only didn't do anything.

**Comment by fgregg** [comment]
and if we try immutable mode, which is how things are opened by `datasette inspect` we duplicate the files!!!

```python
# test_sql_immutable.py
import sqlite3
import sys

db_name = sys.argv[1]
conn = sqlite3.connect(f'file:/app/{db_name}?immutable=1', uri=True)
cur = conn.cursor()
cur.execute('select count(*) from filing')
print(cur.fetchone())
```

**Comment by simonw** [comment]
Oh this is interesting!

Is your hunch here that running this line is causing the file to be stored as a second layer?

https://github.com/simonw/datasette/blob/5aa359b86907d11b3ee601510775a85a90224da8/datasette/utils/__init__.py#L399

I guess it's possible that running a non-read-only query against the database causes one or two bytes to be changed (maybe a transaction ID or similar?)

Modifying the `inspect` command to use `?mode=ro` seems sensible to me. Except.... it should already be opening those files in immutable mode according to this line: https://github.com/simonw/datasette/blob/eff112498ecc499323c26612d707908831446d25/datasette/cli.py#L172-L173

Here's what opening as a `immutables` does:

https://github.com/simonw/datasette/blob/eff112498ecc499323c26612d707908831446d25/datasette/app.py#L258-L260

https://github.com/simonw/datasette/blob/eff112498ecc499323c26612d707908831446d25/datasette/database.py#L98

**Comment by fgregg** [comment]
yes, and i also think that this is causing the apparent memory problems in #1480. when the container starts up, it will make some operation on the database in `immutable` mode which apparently makes some small change to the db file. if that's so, then the db files will be copied to the read/write layer which counts against cloudrun's memory allocation!

**Comment by fgregg** [comment]
when i hack the connect method to open non mutable files with "ro" https://github.com/simonw/datasette/blob/eff112498ecc499323c26612d707908831446d25/datasette/database.py#L79

then: 

```bash
870 B  RUN /bin/sh -c datasette inspect nlrb.db --inspect-file inspect-data.json
```

the `datasette inspect` layer is only the size of the json file!

**Comment by fgregg** [comment]
> yes, and i also think that this is causing the apparent memory problems in #1480. when the container starts up, it will make some operation on the database in `immutable` mode which apparently makes some small change to the db file. if that's so, then the db files will be copied to the read/write layer which counts against cloudrun's memory allocation!
> 
> running a test of that now.

this completely addressed #1480

**Comment by fgregg** [comment]
@simonw, should i open up a new issue for investigating the differences between "immutable=1" and "mode=ro" and possibly switching to "mode=ro". Or would you like to keep that conversation in this issue?

**Comment by simonw** [comment]
Lets switch to `mode=ro` when the `inspect` command runs, we can use this issue for that.

**Comment by fgregg** [comment]
when you are running from docker, you **always** will want to run as `mode=ro` because the same thing that is causing duplication in the inspect layer will cause duplication in the final container read/write layer when `datasette serve` runs.
